### PR TITLE
initial implementation

### PR DESCRIPTION
**Pull request by ekoby**

### CODE REVIEW

File: kx/kx.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+	hash, _ := blake2b.New(SecretKeyBytes, nil)
+	hash.Write(seed)
+	kp.sk = hash.Sum(nil)
+
+	if len(kp.sk) != SecretKeyBytes {
+		return nil, cryptoError
+	}
+
+	kp.pk, err = curve25519.X25519(kp.sk, curve25519.Basepoint)
```
Comment (dovholuknf) [inline kx/kx.go]: why not let this be passed into the function?

File: kx/kx_test.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		return
+	}
+	server_pair, err := NewKeyPair(seedIncrement(seed))
+
+	if err != nil {
+		t.Errorf("failed to get server key pair")
+		return
+	}
+
+    clt_rx, _:= hex.DecodeString("749519c68059bce69f7cfcc7b387a3de1a1e8237d110991323bf62870115731a")
```
Comment (dovholuknf) [inline kx/kx_test.go]: indent nit

File: kx/kx.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+var invalidSeed = errors.New("invalid seed")
+var cryptoError = errors.New("crypto error")
+var notImplemented = errors.New("not implemented")
+
+type KeyPair struct {
+	pk []byte
+	sk []byte
+}
+
+func NewKeyPair(seed []byte) (*KeyPair, error) {
```
Comment (dovholuknf) [inline kx/kx.go]: needs some doc. if nothing else jsut that passing nil is ok and what happens when it's nil

File: sodium.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+type sodiumStream struct {
+	state C.crypto_secretstream_xchacha20poly1305_state
+}
+
+func NewSodiumSendStream(key []byte) (Encryptor,[]byte, error) {
+	res := &sodiumStream{}
+
+	header := C.malloc(C.crypto_secretstream_xchacha20poly1305_HEADERBYTES)
+	rc := C.crypto_secretstream_xchacha20poly1305_init_push(&res.state, (*C.uchar)(header), (*C.uchar)(C.CBytes(key)))
```
Comment (dovholuknf) [inline sodium.go]: does `C.CBytes(key)` not need a corresponding `C.free`?

File: kx/kx.go (hunk 1 above)
Comment (ekoby) [reply kx/kx.go]: can you elaborate? what this and what function?

File: kx/kx_test.go (hunk 2 above)
Comment (ekoby) [reply kx/kx_test.go]: fixed

File: sodium.go (hunk 4 above)
Comment (ekoby) [reply sodium.go]: yes, fixed

File: kx/kx.go (hunk 3 above)
Comment (ekoby) [reply kx/kx.go]: changed to make seeded key generation internal operation, used only for testing

File: kx/kx.go (hunk 1 above)
Comment (dovholuknf) [reply kx/kx.go]: curve25519.X25519 that function. i was asking it i conforms to some interface so that one could change/not use curve25519?

Comment (ekoby) [reply kx/kx.go]: no, and we cannot _not use_ curve25519 because libsodium is using that, key exchange won't work with anything else

**Pull request merged**